internal/pkg/agent/application/upgrade/marker_watcher.go

```diff
@@ -0,0 +1,124 @@
+	Watch() <-chan UpdateMarker
+	Errors() <-chan error
+	Run(ctx context.Context) error
+	Close() error
+}
+
+type MarkerFileWatcher struct {
+	watcher        *fsnotify.Watcher
+	markerFilePath string
+
+	logger *logger.Logger
+
+	updateCh chan UpdateMarker
+	errCh    chan error
+}
+
+func newMarkerFileWatcher(upgradeMarkerFilePath string, logger *logger.Logger) (MarkerWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create upgrade marker watcher: %w", err)
+	} // Watch the upgrade marker file's directory, not the file itself, so we
+
+	// notice the file even if it's deleted and recreated.
+	upgradeMarkerDirPath := filepath.Dir(upgradeMarkerFilePath)
+	err = watcher.Add(upgradeMarkerDirPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set watch on upgrade marker's directory [%s]: %w", upgradeMarkerDirPath, err)
+	}
+
+	return &MarkerFileWatcher{
+		watcher:        watcher,
+		markerFilePath: upgradeMarkerFilePath,
+		logger:         logger,
+		updateCh:       make(chan UpdateMarker),
+		errCh:          make(chan error),
+	}, nil
+}
+
+func (mfw *MarkerFileWatcher) Watch() <-chan UpdateMarker {
+	return mfw.updateCh
+}
+
+func (mfw *MarkerFileWatcher) Errors() <-chan error {
+	return mfw.errCh
+}
+
+func (mfw *MarkerFileWatcher) Run(ctx context.Context) error {
+	// Handle watching
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err, ok := <-mfw.watcher.Errors:
+				mfw.logger.Debug("after there are watch errors")
+				if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+					mfw.logger.Debug("upgrade marker watch's error channel was closed")
+					return
+				}
+				mfw.errCh <- fmt.Errorf("upgrade marker watch returned error: %w", err)
+				continue
+			case e, ok := <-mfw.watcher.Events:
+				mfw.logger.Debugf("after there is a watch event: [%s]", e.String())
+				if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+					mfw.logger.Debug("upgrade marker watch's events channel was closed")
+					return
+				}
+
+				if e.Name != mfw.markerFilePath {
+					// Event is for a file other than the upgrade marker; ignore it.
+					continue
+				}
+
+				switch e.Op {
+				case fsnotify.Create, fsnotify.Write:
+					// Upgrade marker file was created or updated; read its contents
+					// and send them over the update channel.
+
+					marker, err := loadMarker(mfw.markerFilePath)
+					if err != nil {
+						mfw.errCh <- fmt.Errorf("unable to load upgrade marker from watch: %w", err)
+						return
+					}
+
+					mfw.updateCh <- *marker
+				}
+			}
+		}
+	}()
+
+	// Do an initial read from the upgrade marker file, in case the file
+	// is already present before the watching starts.
+	marker, err := loadMarker(mfw.markerFilePath)
+	if err != nil {
+		return fmt.Errorf("unable to load upgrade marker from watch: %w", err)
+	}
+	if marker != nil && marker.Details != nil {
+		mfw.updateCh <- *marker
+	}
+
+	<-ctx.Done()
```

Comment:
we launch a goroutine but then we block here... 
We have the goroutine just to start watching the file as soon as possible while we try and do the first read? 
Couldn't we refactor this to have a single go routine handle the watch and the first read? This would avoid possible races if a write happens between the start of the goroutine and the first read and both try to send the same marker on the update channel

Reply:
I agree with Paolo, if it's gonna block anyway, I don't see why the goroutine is needed.

Reply:
Agreed, I did not implement this well.  Will refactor.  Thanks for calling it out.

Reply:
Refactored in 83577da728.